botocore/credentials.py

```diff
@@ -791,6 +821,157 @@ def _assume_role_base_kwargs(self, config):
+
+    def __init__(self, load_config, client_creator, cache, profile_name,
+                 role_selector=_role_selector,
+                 password_prompter=getpass.getpass,
+                 username_prompter=raw_input):
+        super(AssumeRoleWithSamlProvider, self).__init__(
+            load_config, client_creator, cache, profile_name)
+        self.username_prompter = username_prompter
+        self.password_prompter = password_prompter
+        self.role_selector = role_selector
+
+    def _create_cache_key(self):
+        return self._profile_name.replace(':', '_')
+
+    def _get_role_config_values(self):
+        return self._loaded_config.get('profiles', {}).get(
+            self._profile_name, {})
+
+    def _create_creds_from_response(self, response):
+        config = self._get_role_config_values()
+        if config.get('saml_authentication_type') in ['form']:
+            # If some parameter(s) would require prompting the user for input,
+            # we use a different refresh_func.
+            # We can explore an option in the future to support reprompting for
+            # input, but for now we just error out when the temp creds expire.
+            refresh_func = lambda: exception_raiser(RefreshUnsupportedError)
+        else:
+            refresh_func = lambda: credential_normalizer(
+                self._create_client().assume_role_with_saml(
+                    **self._assume_role_base_kwargs(config))['Credentials'])
+        return RefreshableCredentials(
+            access_key=response['Credentials']['AccessKeyId'],
+            secret_key=response['Credentials']['SecretAccessKey'],
+            token=response['Credentials']['SessionToken'],
+            method=self.METHOD,
+            expiry_time=_parse_if_needed(
+                response['Credentials']['Expiration']),
+            refresh_using=refresh_func)
+
+    def _create_client(self):
+        # sts.assume_role_with_saml() requires no access keys,
+        # but we still need a pair of dummy values here to break recursion.
+        return self._client_creator(
+            'sts', aws_access_key_id='dummy', aws_secret_access_key='dummy')
+
+    def _retrieve_temp_credentials(self):
+        logger.debug("Retrieving credentials via AssumeRoleWithSaml.")
+        response = self._create_client().assume_role_with_saml(
+            **self._assume_role_base_kwargs(self._get_role_config_values()))
+        creds = self._create_creds_from_response(response)
+        return creds, response
+
+    def _assume_role_base_kwargs(self, config):
+        if config.get('saml_authentication_type')=='form':
+            if not config.get('saml_username'):
+                config['saml_username'] = self.username_prompter("Username: ")
+            mapping = {
+                "adfs": {
+                    "username": "ctl00$ContentPlaceHolder1$UsernameTextBox",
+                    "password": "ctl00$ContentPlaceHolder1$PasswordTextBox"},
```

Comment:
Do you mean something like this?

    mapping = {
                "adfs": {
                    "username": "ctl00$ContentPlaceHolder1$UsernameTextBox",
                    "password": "ctl00$ContentPlaceHolder1$PasswordTextBox"},
                "okta": {
                    "username": "username",
                    "password": "password"},
                }.get(config.get('saml_provider'),
                      {"username": "username", "password": "password"})

But I kind of purposely not explicitly give "okta" a name, simply because its field settings are so generic. The point is I don't want in future, when we find another 3rd IdP named "FooBar" who happens to also use the same generic field names "username" and "password", and then we need to configure it as "saml_provider=okta".

We can pull this part of logic out to become a separated function, though.